### PR TITLE
[Logs] Fix: Respect AttributeValueLengthLimit when building otlp LogRecord data

### PR DESCRIPTION
**Pull request by JWilh**
Fixes avoid creating otlp data that is being rejected by monitoring tools, that have strict attribute length and attribute count restrictions.

## Changes
Respect SdkConfiguration.AttributeValueLengthLimit and  SdkConfiguration.AttributeCountLimit when adding attributes to the otlp LogRecord object.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by utpilla** [comment]
@JWilh Thanks for sending this PR! Could you please sign the CLA?

**Comment by JWilh** [comment]
@utpilla Sorry it took a little bit, I was lacking time to find the right person inside my company (right before code freeze and then I was some days off).
It's now properly signed.

**Comment by JWilh** [comment]
@utpilla Does it make sense to add this change to the ChangeLog and does it make sense to add the validation of OTEL_ATTRIBUTE_COUNT_LIMIT also in this pr?

**Comment by JWilh** [comment]
Windows test run failed with 

[xUnit.net 00:00:48.91]     OpenTelemetry.Instrumentation.AspNetCore.Tests.InProcServerTests.ExampleTest [FAIL]
[xUnit.net 00:00:48.91]       System.Net.Http.HttpRequestException : The SSL connection could not be established, see inner exception.
[xUnit.net 00:00:48.92]       ---- System.Security.Authentication.AuthenticationException : The remote certificate is invalid because of errors in the certificate chain: UntrustedRoot

Is this issue already know or just a "transient" error?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:130]: I suggest we get rid of the `AddStringAttribute` method and use `OtlpKeyValueTransformer.Instance.TryTransformTag` even for these one-offs. That way we benefit from the uniform way we process string attributes.

Also, OTLP tag transform has a test suite, so the truncation logic is at least covered there. It'd be nice to also have some tests specifically for the log exporter logic. That said, we do not yet have tests for the trace exporter either with respect to attribute limits.

**Comment by alanwest** [comment]
> Is this issue already know or just a "transient" error?

Not likely related to this PR. I assume it's a flaky test, but would need to investigate.

**Comment by alanwest** [comment]
Related https://github.com/open-telemetry/opentelemetry-specification/pull/2861

**Comment by JWilh** [comment]
> Related [open-telemetry/opentelemetry-specification#2861](https://github.com/open-telemetry/opentelemetry-specification/pull/2861)

I like it, should I already start using the new env var names for the log record limits or better stay with the current ones and they are changed later on?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (JWilh) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs]: I didn't removed the AddStringAttribute function, but replaced the implementation to use the OtlpKeyValueTransformer.Instance.TryTransformTag.
Removing the function itself was for me to excessive (additional if for the TryTransform and for the AttributeCount). 
Is that good for you?

About tests: I will check during the next days when I have a free timeslot. 👍

**Review by alanwest — APPROVED** [review]
LGTM. Thanks for working on this @JWilh!

I'm not inclined to hold this PR up for tests for the log exporter logic as we don't currently have tests for the trace exporter either. But if you have some cycles to add some tests in a follow up that be greatly appreciated! ❤️

**Comment by JWilh** [comment]
Draft for unittest: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3758 
The relevant commit is this one: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3758/commits/44725a645ab957e862ad07c9cea23d6ad8c65a14

**Comment by JWilh** [comment]
@alanwest This pr here is merged by you or @utpilla, is this correct? 
(as I don't have permissions to actually merge it)

**Comment by alanwest** [comment]
> @alanwest This pr here is merged by you or @utpilla, is this correct?

Ah, yes, this is correct! Apologies for not elaborating... we normally wait for at least a couple sets of eyes on a PR before merging. I'll poke one of the other maintainers and see if we can get this merged soon.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (utpilla) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:196]: Do we have to worry about attribute length for `int` values?

Comment (JWilh) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:196]: No, we don't need to worry about it, the tag transformer ignores the length arg in the switch statement for ints.
I just wanted to keep AddStringAttribute and AddIntAttribute consistent and call the TryTransformTag always with the same value length.

**Comment by CodeBlanch** [comment]
Merged so I can get this working with #3760

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (CodeBlanch) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:175]: @alanwest One random thought looking at this PR about how we could improve things...

We build the `OtlpCommon.KeyValue` before we check if we will drop it. Could probably improve that so we skip the allocation(s) if we know we are up against the limit?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (utpilla) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:196]: I think we should remove it from `AddIntAttribute` as it is just not applicable to it. We would never trim or round off numbers.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (JWilh) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:175]: @alanwest I just tried changing it for the LogRecord extension, but we would need something likely a dryrun for TryTransformTag, otherwise we would count attributes as dropped, that were ignored by TryTransformTag before.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs
Comment (JWilh) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/LogRecordExtensions.cs:196]: Addressed in the pr that includes an additional test: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3758

**Pull request merged**
